Random inliner

AndyAyersMS opened the pull request:
This change creates an inline RandomPolicy which assigns a random
benefit to each discretionary inline candidate. Force inlines are
honored as they are with LegacyPolicy. The pool of discretionary
candidates is widened to include the "always inline" cases from the
LegacyPolicy. Discretionary inlines are accepted with a likelihood
that decreases with IL size, in a manner that mirrors the acceptance
probability of the LegacyPolicy.

The RandomPolicy is enabled in DEBUG builds when `JitInlineRandomSeed`
is set to a nonzero seed value. The seed value can be varied to produce
different (but reproducible) random inline patterns.

The random inliner is intended to be used mainly as a stress testing
tool.

AndyAyersMS commented:
@kyulee1 PTAL
cc @dotnet/jit-contrib 

FYI here is a graph of the probability distribution of discretionary inline acceptances from crossgen of mscorlib. For LegacyPolicy anything <= 16 bytes falls into the always inline camp, and anything > 100 is an automatic rejection. The RandomPolicy's threshold settings mimic this distribution in a rough fashion.

![image](https://cloud.githubusercontent.com/assets/10121823/13758732/22419a02-e9e8-11e5-949b-84c8f152edac.png)

sivarv commented on src/jit/inlinepolicy.cpp:
```diff
@@ -26,7 +26,19 @@
+#ifdef DEBUG
+
+    // If JitInlineRandomSeed is set, use the RandomPolicy.
+    unsigned seed = JitConfig.JitInlineRandomSeed();
```
Since Random inline policy is meant as a stress tool, would it make sense to add a new stress mode

STRESS_INLINE_RANDOM?

See compiler.h by searching STRESS_MODES to see the list of existing stress modes. 

In such a case a jistress=1/2 setting would trigger random inline policy for some of the methods. If one wants random policy for all methods we could use

complus_jitstress=1
complus_jitstress_ModeNames=STRESS_INLINE_RANDOM

sivarv replied:
Note that we already have STRESS_INLINE mode that increases the multiplies and the threshold IL size of the method to be considered for in-lining.  But only legacy policy is supporting STRESS_INLINE.  May be we should rename it as STRESS_INLINE_LEGACY since random policy is not affected by it.

AndyAyersMS replied:
I'm actually puzzled why the existing jit stress (at least as far as I know) doesn't have some kind of external seed value. While it is certainly somewhat stressful, it could exhibit far more variation with a combined method hash + seed like I have proposed for the random inliner.

Consider currently that JitStress=2 is still a single run per test case. With an external seed there would be 2^32 variations of JitStress=2 and we could run as little or much of this as we cared.

sivarv replied:
Actually jitstress do support a random seed in the form of

complus_jitstress=any value other than 0 or 2.

See the comment in compStressCompile()

```
    // 0:   No stress (Except when explicitly set in complus_JitStressModeNames)
    // !=2: Vary stress. Performance will be slightly/moderately degraded
    // 2:   Check-all stress. Performance will be REALLY horrible
    stressLevel = getJitStressLevel();
```

And this is how stressLevel is used in determining whether to take 'stress' paths

  ```
  // Get a hash which can be compared with 'weight'

```
    assert(stressArea != 0);
    hash = (info.compMethodHash() ^ stressArea ^ stressLevel) % MAX_STRESS_WEIGHT;

    assert(hash < MAX_STRESS_WEIGHT && weight <= MAX_STRESS_WEIGHT);
    doStress = (hash < weight);
```
```

sivarv replied:
From what i understand, all Jitstress modes are controlled by a single seed value (complus_jistress=value).  It would be a natural use of this seed for random in-liner under stress mode = STRESS_INLINE_RANDOM.  Like all other stress areas, GetPolicy() could call compStressCompile(stressArea, weight) and based on it can choose Random in-lining policy for jit compiling a method.

AndyAyersMS replied:
Interesting. But (as your probably well know) `stressLevel` only influences which stress modes are enabled, not how a particular stress mode behaves when it's enabled.

I suppose I can fold random inlining into the existing jit stress. My main motivation is to assess how likely alternative inlining strategies will uncover issues in the jit, so I'm happy so long as I can run with just random inlining enabled and can also vary the inlining patterns. Tying it into stress and providing an additional seed seems like it would work out ok.

I'm still not sure what the code in `getPolicy` would look like in this case... will work on it a bit.

sivarv commented on src/jit/inlinepolicy.cpp:
```diff
@@ -26,7 +26,19 @@
+    if (seed != 0)
+    {
+        return new (compiler, CMK_Inlining) RandomPolicy(compiler, isPrejitRoot, seed);
+    }
```
How about Random policy by default uses method hash as the seed?  Specifying an additional random seed is optional.  It is an additional control if one wants to use.

AndyAyersMS replied:
Here's what I've come up with:
```
InlinePolicy* InlinePolicy::GetPolicy(Compiler* compiler, bool isPrejitRoot)
{

#ifdef DEBUG

    // Optionally install the RandomPolicy.
    // 
    // We do this when stress is enabled and we're not going to enable
    // legacy inline stress (roughly 50% of the time if both inline
    // stress modes are enabled).
    //
    // Thus we have:
    // All legacy: (default), JitStress=2 with all modes enabled
    //  50/50 mix: JitStress != (0,2) with all modes enabled
    // All random: JitStress != 0 with legacy mode not enabled. Random seed
    //   is the jit stress value

    bool useRandomPolicy = compiler->compInlineRandomStress() && !compiler->compInlineStress();

    if (useRandomPolicy)
    {
        unsigned seed = getJitStressLevel();
        assert(seed != 0);
        return new (compiler, CMK_Inlining) RandomPolicy(compiler, isPrejitRoot, seed);
    }

#endif // DEBUG
```

That way I can get my "all random w/varying seed" and we mix in some random behavior during JitStress with values other than 2.

sivarv replied:
What is not clear to me is under what conditions/settings these one of these routines return true?
compInlineRandomStress()
compInlineStress()

My suggestion is to call compStressCompile(stressarea, 50) with a weight of 50% (or 0.5 probability) and select random policy.  If either no stress (jistress=0) or the probability doesn't favor, default is to choose legacy policy.
If random policy gets selected, then it can further use jitstress level as the seed.

Jitstress=0: no stress, always uses legacy policy
jitstress=2: one of random or legacy policy could be used based on method hash.  Further '2' is used as the random seed by random policy.
Jitstress= value other than 0 or 2: Same as above.

AndyAyersMS replied:
It's set up this way because I would like there to be a mode where RandomPolicy is the only policy used, and I can control the seed. The missing functions are:
```
compInlineRandomStress == compStressCompile(STRESS_RANDOM_INLINE, 100)
compInlineStress = compStressCompille(STRESS_INLINE, 50)
```
Thus by setting things as follows I can get the mode I want:
```
set complus_jitstress=2020202
set complus_jitstressModeNames=STRESS_RANDOM_INLINE
set complus_jitstressModeNamesNot=STRESS_INLINE
```
(note setting `jitStressModeNames` does not exclude other modes from being active, which seems a little odd).

sivarv replied:
I think my suggestion is same as what you propose to do with a very minor difference:

With your suggestion:
compInlineRandomStress = true with 100% probability (because MAX_STRESS_WEIGHT is 100).
compInlineStress = true with 0.5 probability

since the condition to select random policy is

compiler->compInlineRandomStress() && !compiler->compInlineStress()

essentially with 0.5 probability random policy gets selected.

Instead my suggestion is

```
useRandomPolicy = compStressCompile(STRESS_RANDOM_INLINE, 50);
if (useRandomPolicy)
{
// use random policy
}
else
{
   // use legacy policy
}

```
Note that random policy will remain in effect for all methods with the following settings alone with my suggestion:

```
set complus_jitstress=2020202
set complus_jitstressModeNames=STRESS_RANDOM_INLINE
```

Because JitSTressModeNames settings will make compStressCompile() return true irrespective of the weight parameter passed to it.

AndyAyersMS replied:
Either way the remainder of the stress modes would still be active, right? 

What I was hoping to do with the RandomPolicy was to assess how robust the rest of the jit is to changes in the inliner (since I plan to make changes). Ideally this means just inlining differently but "plausibly" (at least as far as not creating crazily big methods like STRESS_INLINE does). I guess I can get the more or less the same information even if enabling the RandomPolicy also enables a bunch of non-inline stress but this is not what I'd envisioned.

It does not seem easy right now to just enable one stress mode.

sivarv replied:
Yes, either way remaining stress modes will be in effect.

To just run with one stress mode enabled, we would have to use the following settings (I know it is a bit clumsy)

complus_jitstress=value
complus_JitStressModeNames=STRESS_RANDOM_INLINE
complus_JitStressModeNamesNot="delimited list of all stress mode names listed in compiler.h other than RANDOM_INLINE"

To make it easier, we can think of adding a new config option (say complus_JitStressModeNamesOnly) to exercise only the stress modes specified with complus_JitStressModeNames setting.

AndyAyersMS replied:
Ok, let me work on this and push an update.

AndyAyersMS commented:
@dotnet-bot retest Windows_NT x86 Release Legacy Backend Build and Test

kyulee1 commented:
Overall looks good to me. I like the graph shows this random policy is not really random but a reasonably random that follows the legacy policy trajectory to reasonably broaden the test coverage. 
I just noticed that the private functions ```SetCandidate/SetFailure/SetNever``` seem identical for both legacy/random policy, so you might consider moving them to base InlinePolicy. But it's up to you since you might alter this later.

AndyAyersMS commented:
@sivarv Updated to integrate with stress, and allow the *only* mode.

@kyulee1 you're right that there is some duplication between `LegacyPolicy` and `RandomPolicy`. I'll refator this down once I've introduced another policy or two and can see the parts that are common to all.

Since I've changed the behavior of JitStress I might as well see if anything new turns up.

@dotnet-bot test Windows_NT x64 jitstress2

sivarv commented on src/jit/compiler.cpp:
```diff
@@ -2971,15 +2975,28 @@ bool            Compiler::compStressCompile(compStressArea  stressArea,
     strStressModeNames = JitConfig.JitStressModeNames();
-    if ((strStressModeNames != NULL) &&
-        (wcsstr(strStressModeNames, s_compStressModeNames[stressArea]) != NULL))
+    if (strStressModeNames != NULL)
```
(Minor) CC: nullptr

sivarv commented on src/jit/compiler.cpp:
```diff
@@ -2971,15 +2975,28 @@ bool            Compiler::compStressCompile(compStressArea  stressArea,
+    if (strStressModeNames != NULL)
     {
-        if (verbose)
+        if (wcsstr(strStressModeNames, s_compStressModeNames[stressArea]) != NULL)
```
(Minor) CC: nullptr

sivarv commented on src/jit/compiler.h:
```diff
@@ -7607,7 +7607,7 @@ public :
                                                                                                 \
         STRESS_MODE(REGS) STRESS_MODE(DBL_ALN) STRESS_MODE(LCL_FLDS) STRESS_MODE(UNROLL_LOOPS)  \
-        STRESS_MODE(MAKE_CSE) STRESS_MODE(INLINE) STRESS_MODE(CLONE_EXPR)                       \
+        STRESS_MODE(MAKE_CSE) STRESS_MODE(LEGACY_INLINE) STRESS_MODE(CLONE_EXPR)                \
```
I see two references to preprocessor constant STRESS_INLINE.  Post renaming it as STRESS_LEGACY_INLINE don't we need to rename in importer.cpp as well?  Looks like you might have forgotten to include/push importer changes.

AndyAyersMS replied:
I think you may be looking at an older version of that file. I removed the importer.cpp reference in 97c0fcb.

sivarv replied:
Looks like i was looking at a bit of old source code.  Please ignore above comment.

sivarv commented:
Looks good.

mmitche commented:
@dotnet-bot test Windows_NT jitstress2

sejongoh commented:
@dotnet-bot test Windows_NT jitstress2

AndyAyersMS commented:
Am going to look at the stress failures.

AndyAyersMS commented:
The stress failures are a pre-existing issue. Opened #3760 for that and will fix separately. Going to merge since the regular CI results are green.
